### ISSUE TITLE
Allow ability to call `predict(mach, *)` on a task (as well as new input data)?

### ISSUE THREAD

**Issue by ablaom** [alan-turing-institute/MLJ.jl#158]
I'm happy with the status quo, but this is being attempted. See #156.

What do others think?

Also, do we allow loss functions to be called on tasks? (also attempted)

**Comment by fkiraly** [comment]
I'm all in favour of dispatch-on-task!
Especially since it seems to me the only general way forward for non-tabular learning.

I may be biased though...
https://github.com/alan-turing-institute/sktime/issues/1
(part 4)

**Comment by fkiraly** [comment]
Loss function on task though seems to make less sense to me.

A loss function needs predictions and actual values, the but task doesn't contain these, at least not predictions?

I'd rather see the natural place of loss functions as arguments of `evaluate`, or of a yet-to-be-written benchmark orchestrator.

**Comment by davidbp** [comment]
In the documentation it is stated  "**In MLJ one can either wrap data** for supervised learning in a formal task (see Working with Tasks), or **work directly with the data**, split into its input and target parts:"
    
If in MLJ  one can either work with tasks/data I think it is perfectly reasonable to allow  calling both `fit!` and `predict` with either of those options. Why `fit!(model, some_task)` is OK but `predict(model, some_task)` is not?

**Comment by fkiraly** [comment]
@davidbp very interesting, we had extensive design discussion at sktime and with the mlj team:
https://github.com/alan-turing-institute/sktime/issues/20

In terms of ontology, I think there's a clear distinction between two kinds of tasks:
(i) "task to solve & evaluate": the "kaggle style task" that you seem to have in mind: provides a full specification of how to evaluate a strategy, it is hence a "strategy-shaped hole" (co-strategy pattern) with respect to benchmarking.
(ii) "task to solve": just a specification of what the strategy should do with the data: specification of how to apply it to data, i.e., a "strategy-shaled hole" (co-strategy pattern) with respect to fit/predict.

In our internal discussion, there was a lot of confusion between the two, and disagreement regarding whether
(a) one should have classes/objects for both or just one
(b) whether the data should be part of it, whether a re-sampling strategy for evaluation should be part of it, whether a loss function should be part of it

Regarding the aspect of loss function: one could argue that it shouldn't be part, since you can measure the quality of an algorithm in multiple respects. It makes more sense in (i) than in (ii) though, since it would allow you to specify a full benchmarking experiment.

In addition, please note: the loss function used in evaluation does not necessarily need to concide with a loss function used in training/tuning - especially since a model may not be fitted by minimizing a loss function!

**Comment by fkiraly** [comment]
addendum: actually, I don't think that anyone suggested that loss functions should be part of the type (i) task, called option 3 in https://github.com/alan-turing-institute/sktime/issues/20 , so this is a new idea of @davidbp .

**Comment by ablaom** [comment]
Okay, sorry. On second thoughts calling `predict(mach, task)` is a bit silly because `mach` already wraps data (from a task, if a task was used to construct the machine). I think what @davidbp wants is to be able to call `predict(mach, rows=10:23)` or whatever (just like you can do `fit!(mach, rows=200:300)`) and in fact this is already implemented (although not documented, sorry).

Note that `predict(mach, rows=...)` will not work if mach is a `NodalMachine` because in that case `predict` doesn't actually predict but rather defines a new node  with operation `predict` and this constructor must be provided an `X` representing the node where the operation looks for its input data.
